src/Cli/dotnet/commands/dotnet-run/RunCommand.cs

```diff
@@ -201,77 +212,128 @@ private List<string> GetRestoreArguments()
 
-            string runProgram = project.GetPropertyValue("RunCommand");
-            if (string.IsNullOrEmpty(runProgram))
-            {
-                ThrowUnableToRunError(project);
+                if (applicationArgs.Any())
+                {
+                    runArguments += " " + ArgumentEscaper.EscapeAndConcatenateArgArrayForProcessStart(applicationArgs);
+                }
+                return new(runProgram, runArguments, runWorkingDirectory);
             }
 
-            string runArguments = project.GetPropertyValue("RunArguments");
-            string runWorkingDirectory = project.GetPropertyValue("RunWorkingDirectory");
-
-            if (Args.Any())
+            static ICommand CreateCommandFromRunProperties(ProjectInstance project, RunProperties runProperties)
             {
-                runArguments += " " + ArgumentEscaper.EscapeAndConcatenateArgArrayForProcessStart(Args);
-            }
+                CommandSpec commandSpec = new(runProperties.RunCommand, runProperties.RunArguments);
 
-            CommandSpec commandSpec = new(runProgram, runArguments);
+                var command = CommandFactoryUsingResolver.Create(commandSpec)
+                    .WorkingDirectory(runProperties.RunWorkingDirectory);
 
-            var command = CommandFactoryUsingResolver.Create(commandSpec)
-                .WorkingDirectory(runWorkingDirectory);
+                var rootVariableName = EnvironmentVariableNames.TryGetDotNetRootVariableName(
+                    project.GetPropertyValue("RuntimeIdentifier"),
+                    project.GetPropertyValue("DefaultAppHostRuntimeIdentifier"),
+                    project.GetPropertyValue("TargetFrameworkVersion"));
 
-            var rootVariableName = EnvironmentVariableNames.TryGetDotNetRootVariableName(
-                project.GetPropertyValue("RuntimeIdentifier"),
-                project.GetPropertyValue("DefaultAppHostRuntimeIdentifier"),
-                project.GetPropertyValue("TargetFrameworkVersion"));
+                if (rootVariableName != null && Environment.GetEnvironmentVariable(rootVariableName) == null)
+                {
+                    command.EnvironmentVariable(rootVariableName, Path.GetDirectoryName(new Muxer().MuxerPath));
+                }
+                return command;
+            }
 
-            if (rootVariableName != null && Environment.GetEnvironmentVariable(rootVariableName) == null)
+            static void InvokeRunArgumentsTarget(ProjectInstance project)
             {
-                command.EnvironmentVariable(rootVariableName, Path.GetDirectoryName(new Muxer().MuxerPath));
-            }
+                if (project.Build(["ComputeRunArguments"], loggers: [new BinaryLogger { Parameters = "{}.binlog" }], remoteLoggers: null, out var _targetOutputs))
+                {
 
-            return command;
+                }
+                else
+                {
+                    throw new GracefulException("boom");
```

Comment:
nit: need a localized message here. We also need a way to surface errors from the build here - we need to hook up _some_ kind of logger, and that logger should probably be TL, but we can't easily do this from code. cc @rainersigwald for this gap - maybe we need to make TL a public API?

Reply:
You would need to access the current logger, not hook up a new one, right? Since it already built? So would we need to use a LoggingService or a ProjectCollection to retain access to it?

If you have a list of the ILoggers MSBuild used, I don't think MSBuild needs to know the ILogger is a TL to log a message, but I also don't know how we'd log it from there.

Reply:
@Forgind part of the problem is that we're in a hybrid mode of using MSBuild here:

* the restore/build commands that `run` inherently relies on use MSBuild like a CLI app, so they express opinions about logging in terms of the CLI arguments passed to MSBuild, not anything more structured
* this evaluation/target execution for `run` uses MSBuild via the API, so if we want to keep things unified in terms of experience we should try to recreate similar logger configurations, but because we've delegated this to MSBuild via CLI arguments it's hard to get that level of consistency

Reply:
Rainer blessed the temporary reflection I've used. Short term will will IVT MSBuild so that the CLI can access the TL types, but medium/long term MSbuild should refactor argument parsing to be something that callers can programmatically invoke and get structured outputs for.